File: scripts/tools/repo-hygiene-check.mjs
#!/usr/bin/env node
import { existsSync } from "node:fs";
import { spawn } from "node:child_process";
import path from "node:path";
import { fileURLToPath } from "node:url";

const ROOT_DIR = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..", "..");
const HIGH_RISK_PREFIXES = ["apps/", "clients/", "packages/", "deploy/", "scripts/", "docs/"];
const HIGH_RISK_FILES = new Set(["package.json", "pnpm-lock.yaml", "pnpm-workspace.yaml", "tsconfig.base.json"]);
const RUNTIME_PATHS = ["runtime", "logs", "home-ktv-media", "songs-sample", ".worktrees", "worktrees"];

if (import.meta.url === `file://${process.argv[1]}`) {
  main(process.argv.slice(2)).catch((error) => {
    console.error(error instanceof Error ? error.message : String(error));
    process.exitCode = 1;
  });
}

export async function main(argv, dependencies = {}) {
  const options = parseArgs(argv);
  if (options.help) {
    printUsage();
    return;
  }

  const report = await buildHygieneReport(options, dependencies);
  if (options.json) {
    console.log(JSON.stringify(report, null, 2));
  } else {
    printReport(report);
  }

  if (options.failOnDirty && (report.summary.trackedDirty > 0 || report.summary.highRiskUntracked > 0)) {
    process.exitCode = 1;
  }
}

export async function buildHygieneReport(options = {}, dependencies = {}) {
  const rootDir = options.rootDir || ROOT_DIR;
  const runGit = dependencies.runGit ?? ((args) => run("git", args, rootDir));
  const pathExists = dependencies.pathExists ?? ((filePath) => existsSync(path.join(rootDir, filePath)));

  const [status, branch] = await Promise.all([
    runGit(["status", "--porcelain=v1", "-uall"]),
    runGit(["status", "--short", "--branch"])
  ]);

  if (status.code !== 0) {
    throw new Error(status.stderr || "git status failed");
  }

  const entries = parsePorcelainStatus(status.stdout);
  const trackedDirty = entries.filter((entry) => entry.kind !== "untracked");
  const untracked = entries.filter((entry) => entry.kind === "untracked");
  const highRiskUntracked = untracked.filter((entry) => isHighRiskPath(entry.path));
  const runtimePaths = RUNTIME_PATHS.filter((name) => pathExists(name));

  return {
    branch: parseBranchSummary(branch.stdout),
    checkedAt: new Date().toISOString(),
    rootDir,
    summary: {
      highRiskUntracked: highRiskUntracked.length,
      runtimePaths: runtimePaths.length,
      trackedDirty: trackedDirty.length,
      untracked: untracked.length
    },
    trackedDirty,
    untrackedHighRisk: highRiskUntracked,
    runtimePaths
  };
}

export function parseArgs(argv) {
  const options = {
    failOnDirty: false,
    help: false,
    json: false,
    rootDir: ""
  };

  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === "--") {
      continue;
    } else if (arg === "--help" || arg === "-h") {
      options.help = true;
    } else if (arg === "--json") {
      options.json = true;
    } else if (arg === "--fail-on-dirty") {
      options.failOnDirty = true;
    } else if (arg === "--root") {
      options.rootDir = readOptionValue(argv, ++index, "--root");
    } else {
      throw new Error(`Unknown option: ${arg}`);
    }
  }

  return options;
}

export function parsePorcelainStatus(stdout) {
  return stdout
    .split(/\r?\n/u)
    .map((line) => line.trimEnd())
    .filter(Boolean)
    .map((line) => {
      const code = line.slice(0, 2);
      const rawPath = line.slice(3);
      return {
        code,
        kind: code === "??" ? "untracked" : "tracked",
        path: normalizeStatusPath(rawPath)
      };
    });
}

export function isHighRiskPath(filePath) {
  return HIGH_RISK_FILES.has(filePath) || HIGH_RISK_PREFIXES.some((prefix) => filePath.startsWith(prefix));
}

function parseBranchSummary(stdout) {
  const firstLine = stdout.split(/\r?\n/u).find(Boolean) || "";
  return firstLine.replace(/^##\s*/u, "");
}

function normalizeStatusPath(filePath) {
  const renameSeparator = " -> ";
  return filePath.includes(renameSeparator) ? filePath.split(renameSeparator).at(-1) || filePath : filePath;
}

function printReport(report) {
  console.log("HomeKTV repo hygiene");
  console.log(`Root: ${report.rootDir}`);
  console.log(`Branch: ${report.branch || "(unknown)"}`);
  console.log(
    `Summary: tracked dirty ${report.summary.trackedDirty}, untracked ${report.summary.untracked}, high-risk untracked ${report.summary.highRiskUntracked}`
  );
  console.log("");

  if (report.trackedDirty.length > 0) {
    console.log("Tracked dirty files:");
    for (const entry of report.trackedDirty) {
      console.log(`  ${entry.code} ${entry.path}`);
    }
    console.log("");
  }

  if (report.untrackedHighRisk.length > 0) {
    console.log("High-risk untracked files:");
    for (const entry of report.untrackedHighRisk) {
      console.log(`  ${entry.path}`);
    }
    console.log("");
  }

  if (report.runtimePaths.length > 0) {
    console.log("Ignored/runtime paths present:");
    for (const filePath of report.runtimePaths) {
      console.log(`  ${filePath}/`);
    }
    console.log("");
  }

  if (report.summary.trackedDirty === 0 && report.summary.highRiskUntracked === 0) {
    console.log("No tracked dirty files or high-risk untracked files detected.");
  }
}

function printUsage() {
  console.log(
    [
      "Usage: node scripts/tools/repo-hygiene-check.mjs [options]",
      "",
      "Options:",
      "  --json             Print JSON report",
      "  --fail-on-dirty    Exit 1 when tracked or high-risk untracked files exist",
      "  --root <path>      Git repository root, default: current project",
      "  -h, --help         Show help"
    ].join("\n")
  );
}

function run(program, args, cwd) {
  return new Promise((resolve) => {
    const child = spawn(program, args, {
      cwd
    });
    let stdout = "";
    let stderr = "";
    child.stdout.on("data", (chunk) => {
      stdout += chunk;
    });
    child.stderr.on("data", (chunk) => {
      stderr += chunk;
    });
    child.on("close", (code) => {
      resolve({ code: code ?? 0, stderr, stdout });
    });
  });
}

function readOptionValue(argv, index, name) {
  const value = argv[index];
  if (!value) {
    throw new Error(`${name} requires a value`);
  }
  return value;
}

File: scripts/tools/repo_hygiene_check.py
#!/usr/bin/env python3
"""Check HomeKTV repository hygiene before commits."""

from __future__ import annotations

import argparse
import json
import subprocess
import sys
from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from pathlib import Path


REPO_ROOT = Path(__file__).resolve().parents[2]
HIGH_RISK_PREFIXES = ("apps/", "clients/", "packages/", "deploy/", "scripts/", "docs/")
HIGH_RISK_FILES = {"package.json", "pnpm-lock.yaml", "pnpm-workspace.yaml", "tsconfig.base.json"}
RUNTIME_PATHS = ("runtime", "logs", "home-ktv-media", "songs-sample", ".codex", ".planning/reports", ".worktrees", "worktrees")


@dataclass(frozen=True)
class StatusEntry:
    code: str
    kind: str
    path: str


def main(argv: list[str] | None = None) -> int:
    args = parse_args(argv)
    report = build_hygiene_report(args.root)
    if args.json:
        print(json.dumps(report, ensure_ascii=False, indent=2))
    else:
        print_report(report)

    summary = report["summary"]
    if args.fail_on_dirty and (summary["trackedDirty"] > 0 or summary["highRiskUntracked"] > 0):
        return 1
    return 0


def parse_args(argv: list[str] | None = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--json", action="store_true", help="Print JSON report.")
    parser.add_argument("--fail-on-dirty", action="store_true", help="Exit 1 when tracked or high-risk untracked files exist.")
    parser.add_argument("--root", type=Path, default=REPO_ROOT, help="Git repository root.")
    raw_args = sys.argv[1:] if argv is None else argv
    forwarded_args = [arg for arg in raw_args if arg != "--"]
    return parser.parse_args(forwarded_args)


def build_hygiene_report(root_dir: Path | str = REPO_ROOT) -> dict:
    root = Path(root_dir).resolve()
    status = run_git(root, "status", "--porcelain=v1", "-uall")
    branch = run_git(root, "status", "--short", "--branch")
    entries = parse_porcelain_status(status.stdout)
    tracked_dirty = [entry for entry in entries if entry.kind != "untracked"]
    untracked = [entry for entry in entries if entry.kind == "untracked"]
    high_risk_untracked = [entry for entry in untracked if is_high_risk_path(entry.path)]
    runtime_paths = [name for name in RUNTIME_PATHS if (root / name).exists()]

    return {
        "branch": parse_branch_summary(branch.stdout),
        "checkedAt": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "rootDir": str(root),
        "summary": {
            "highRiskUntracked": len(high_risk_untracked),
            "runtimePaths": len(runtime_paths),
            "trackedDirty": len(tracked_dirty),
            "untracked": len(untracked),
        },
        "trackedDirty": [asdict(entry) for entry in tracked_dirty],
        "untrackedHighRisk": [asdict(entry) for entry in high_risk_untracked],
        "runtimePaths": runtime_paths,
    }


def parse_porcelain_status(stdout: str) -> list[StatusEntry]:
    entries: list[StatusEntry] = []
    for raw_line in stdout.splitlines():
        line = raw_line.rstrip()
        if not line:
            continue
        code = line[:2]
        raw_path = line[3:]
        entries.append(StatusEntry(code=code, kind="untracked" if code == "??" else "tracked", path=normalize_status_path(raw_path)))
    return entries


def is_high_risk_path(file_path: str) -> bool:
    return file_path in HIGH_RISK_FILES or file_path.startswith(HIGH_RISK_PREFIXES)


def parse_branch_summary(stdout: str) -> str:
    for line in stdout.splitlines():
        if line:
            return line.removeprefix("## ").strip()
    return ""


def normalize_status_path(file_path: str) -> str:
    return file_path.split(" -> ")[-1] if " -> " in file_path else file_path


def print_report(report: dict) -> None:
    print("HomeKTV repo hygiene")
    print(f"Root: {report['rootDir']}")
    print(f"Branch: {report['branch'] or '(unknown)'}")
    summary = report["summary"]
    print(
        "Summary: tracked dirty "
        f"{summary['trackedDirty']}, untracked {summary['untracked']}, high-risk untracked {summary['highRiskUntracked']}"
    )
    print()

    if report["trackedDirty"]:
        print("Tracked dirty files:")
        for entry in report["trackedDirty"]:
            print(f"  {entry['code']} {entry['path']}")
        print()

    if report["untrackedHighRisk"]:
        print("High-risk untracked files:")
        for entry in report["untrackedHighRisk"]:
            print(f"  {entry['path']}")
        print()

    if report["runtimePaths"]:
        print("Ignored/runtime paths present:")
        for file_path in report["runtimePaths"]:
            print(f"  {file_path}/")
        print()

    if summary["trackedDirty"] == 0 and summary["highRiskUntracked"] == 0:
        print("No tracked dirty files or high-risk untracked files detected.")


def run_git(root: Path, *args: str) -> subprocess.CompletedProcess[str]:
    result = subprocess.run(["git", *args], cwd=root, check=False, text=True, capture_output=True)
    if result.returncode != 0:
        raise RuntimeError(result.stderr.strip() or "git status failed")
    return result


if __name__ == "__main__":
    raise SystemExit(main())
